### PR TITLE
[WIP] DX 26172: Add RocksDB as backing store to Nessie

### PR DESCRIPTION
**Pull request by stevelorddremio**

### CODE REVIEW

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import com.dremio.nessie.versioned.impl.InternalRef;
+import com.dremio.nessie.versioned.impl.condition.ConditionExpression;
+import com.dremio.nessie.versioned.impl.condition.UpdateExpression;
+import com.dremio.nessie.versioned.store.Id;
+import com.dremio.nessie.versioned.store.LoadStep;
+import com.dremio.nessie.versioned.store.SaveOp;
+import com.dremio.nessie.versioned.store.Store;
+import com.dremio.nessie.versioned.store.ValueType;
+
+public class RocksDBStore implements Store {
```
Comment (stevelorddremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java]: Store should in fact be implemented in RocksDBStoreManager, not RocksDBStore. It is following the implementation of Dremio more closely than dynamodb in nessie.

Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java]: This is now fixed by moving implements.

File: pom.xml [hunk 2]
```diff
@@ -139,6 +139,7 @@
     <logback.version>1.2.3</logback.version>
     <protobuf.version>3.13.0</protobuf.version>
     <prometheus.version>0.9.0</prometheus.version>
+    <rocksdb.version>5.14.2</rocksdb.version>
```
Comment (kylep-dremio) [inline pom.xml:142]: Any reason not to use one of the latest?

File: servers/quarkus-server/src/main/resources/application.properties [hunk 3]
```diff
@@ -28,7 +28,7 @@ nessie.version.store.jgit.type=DISK
 ### where to put your git repository. Only used if nessie.version.store.jgit.type=JGIT
 nessie.version.store.jgit.directory=/tmp/jgit
 ### backing store for jgit dynamo version store. Either INMEMORY or DYNAMO. Typically DYNAMO is used unless performing a specific test
-nessie.backends.type=DYNAMO
```
Comment (kylep-dremio) [inline servers/quarkus-server/src/main/resources/application.properties:31]: Why has this changed?

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java:1]: I believe we talked about creating another project that had a dependency back to the Dynamo project to get access to the relevant interfaces that had not yet moved. That way you would simply have to change the project to wherever the interfaces ended up. It seems everything is created in the Dynamo project currently.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import org.rocksdb.ColumnFamilyDescriptor;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.FlushOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class RocksDBStore implements AutoCloseable {
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java:31]: Please add JavaDocs.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+   */
+  public RocksDB openDB(final DBOptions dboptions, final String path, final List<ColumnFamilyDescriptor> columnNames,
+                        List<ColumnFamilyHandle> familyHandles) throws RocksDBException {
+    boolean printLockMessage = true;
+
+    while (true) {
+      try {
+        return RocksDB.open(dboptions, path, columnNames, familyHandles);
+      } catch (RocksDBException e) {
+        if (e.getStatus().getCode() != Status.Code.IOError || !e.getStatus().getState().contains("While lock")) {
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:201]: Please brace the individual conditions.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+          throw e;
+        }
+
+        // Don't retry if request came from a CLI command (noDBOpenRetry = true)
+        if (noDBOpenRetry) {
+          LOGGER.error("Lock file to RocksDB is currently held by another process.  Stop other process before retrying.");
+          System.out.println("Lock file to RocksDB is currently held by another process.  Stop other process before retrying.");
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:208]: This seems problematic to log to sout....is that how it happens in the rest of the codebase?

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+  public static final byte[] REF_COLUMN_FAMILY = "ref".getBytes();
+  public static final byte[] L1_COLUMN_FAMILY = "l1".getBytes();
+  public static final byte[] L2_COLUMN_FAMILY = "l2".getBytes();
+  public static final byte[] L3_COLUMN_FAMILY = "l3".getBytes();
+  public static final byte[] VALUE_COLUMN_FAMILY = "value".getBytes();
+  public static final byte[] KEY_FRAGMENT_COLUMN_FAMILY = "keyfragment".getBytes();
+  public static final byte[] COMMIT_METADATA_AUTHOR_COLUMN_FAMILY = "commitmetadataauthor".getBytes();
+  public static final byte[] COMMIT_METADATA_MESSAGE_COLUMN_FAMILY = "commitmetadatamessage".getBytes();
+  private static ArrayList<byte[]> COLUMN_FAMILIES = new ArrayList<byte[]>() {{
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:67]: Should store List<> instead of ArrayList - always have the variables be the most generic level of interface possible.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    List<byte[]> families = null;
+
+    //TODO implement valiation of path existing
+
+    try (final Options options = new Options()) {
+      options.setCreateIfMissing(true);
+      // get a list of existing families.
+      families = new ArrayList<>(RocksDB.listColumnFamilies(options, baseDirectory));
+    } catch (RocksDBException e) {
+      LOGGER.error("Error extracting RocksDB options", e);
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:145]: Seems like you should be propagating the exceptions?

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+      }
+    }
+    // populate the local cache with the existing tables.
+    for (int i = 0; i < families.size(); i++) {
+      byte[] family = families.get(i);
+      if (Arrays.equals(family, RocksDB.DEFAULT_COLUMN_FAMILY)) {
+        defaultHandle = familyHandles.get(i);
+      } else {
+        String name = new String(family, UTF_8);
+        final ColumnFamilyHandle handle = familyHandles.get(i);
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:175]: Where does familyHandles have values added to it?

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/store/dynamo/DynamoGeneralReadFailure.java [hunk 11]
```diff
@@ -15,7 +15,7 @@
  */
 package com.dremio.nessie.versioned.store.dynamo;
 
-class DynamoGeneralReadFailure extends RuntimeException {
+public class DynamoGeneralReadFailure extends RuntimeException {
```
Comment (kylep-dremio) [inline versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/store/dynamo/DynamoGeneralReadFailure.java:18]: Why this change?

File: pom.xml (hunk 2 above)
Comment (stevelorddremio) [reply pom.xml:142]: I was in favour of using latest, particularly at this is so far behind. Can't remember who told me, but they advised using same version as Dremio uses. Latest is 6.13.3. Should be a small matter to upgrade.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java (hunk 7 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:208]: Oops. No, this came from the sample code. I'll use LOGGER.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java (hunk 8 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:67]: Agreed. Good catch, I'll update.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java (hunk 9 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:145]: Agreed. Will change.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java (hunk 10 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:175]: in start(), line 150. These are defined in COLUMN_FAMILIES.

Comment (kylep-dremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStoreManager.java:175]: That adds to families, not familyHandles. familyHandles is created on line 154...

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/store/dynamo/DynamoGeneralReadFailure.java (hunk 11 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/store/dynamo/DynamoGeneralReadFailure.java:18]: Not sure. I think my IntelliJ may have complained. I'll revert and check it still builds.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java (hunk 4 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java:1]: Ah, yes. I went with different sub-package. By project do you mean Maven module, ie outside of nessie-versioned, or totally outside of Nessie, repo?

File: servers/quarkus-server/src/main/resources/application.properties (hunk 3 above)
Comment (stevelorddremio) [reply servers/quarkus-server/src/main/resources/application.properties:31]: I changed it to INMEMORY to get it working initially. I'll revert that.

File: versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java (hunk 4 above)
Comment (stevelorddremio) [reply versioned/dynamodb/src/main/java/com/dremio/nessie/versioned/implrocksdb/RocksDBStore.java:1]: Agreed. This should reside in package com.dremio.nessie.versioned.store.rocksdb.

**Comment by kylep-dremio** [comment]
Closing as outdated.
